### PR TITLE
GH-32338: [C++] Add IPC support for Run-End Encoded Arrays 

### LINKED ISSUE #32338: [Format][C++] Support RLE in IPC format
**Reporter**: [Tobias Zagorni](https://issues.apache.org/jira/browse/ARROW-17027) / @zagto
#### PRs and other links:
- [GitHub Pull Request #14340](https://github.com/apache/arrow/pull/14340)

<sub>**Note**: *This issue was originally created as [ARROW-17027](https://issues.apache.org/jira/browse/ARROW-17027). Please see the [migration documentation](https://github.com/apache/arrow/issues/14542) for further details.*</sub>

### PR DESCRIPTION
**Pull request by felipecrv**
* Closes #14340
 * Closes #32773
 * Closes #14179
 
* Closes: #32338

### CODE REVIEW

**Comment by felipecrv** [comment]
Commits with integration tests enabled are on the way.

**Comment by felipecrv** [comment]
@zeroshade

**Review by zeroshade — COMMENTED** [review]
Overall LGTM just a few questions

File: cpp/src/arrow/array/array_run_end.cc [hunk 1]
```diff
@@ -85,6 +86,53 @@ void RunEndEncodedArray::SetData(const std::shared_ptr<ArrayData>& data) {
...
+
+  switch (run_ends_array_->type_id()) {
+    case Type::INT16:
+      return MakeLogicalRunEnds<Int16Type>(*this, physical_offset, physical_length);
+    case Type::INT32:
+      return MakeLogicalRunEnds<Int32Type>(*this, physical_offset, physical_length);
+    default:
+      break;
+  }
+  return MakeLogicalRunEnds<Int64Type>(*this, physical_offset, physical_length);
```
[the remark this answers is not in the document]
Comment (felipecrv) [reply cpp/src/arrow/array/array_run_end.cc:127]: In other projects, I've been annoyed by MSVC complaining the function doesn't return a value when all the returns were in the `switch`, but I can change it here if Arrows MSVC warnings aren't so aggressive.

File: cpp/src/arrow/array/array_run_end.cc [hunk 2]
```diff
@@ -85,6 +86,53 @@ void RunEndEncodedArray::SetData(const std::shared_ptr<ArrayData>& data) {
...
+  if (physical_length > 0) {
+    for (int64_t i = 0; i < physical_length - 1; i++) {
+      const auto run_end = run_ends[physical_offset + i] - self.offset();
+      DCHECK_LT(run_end, self.length());
+      RETURN_NOT_OK(builder.Append(static_cast<RunEndCType>(run_end)));
+    }
+    DCHECK_GE(run_ends[physical_offset + physical_length - 1] - self.offset(),
+              self.length());
+    RETURN_NOT_OK(builder.Append(static_cast<RunEndCType>(self.length())));
+  }
```
[the remark this answers is not in the document]
Comment (felipecrv) [reply cpp/src/arrow/array/array_run_end.cc:108]: Sure. I've forgot. But it's good that I've got a full CI run based on this version which will become a fallback when I add the if.

File: cpp/src/arrow/array/array_run_end.h [hunk 3]
```diff
@@ -87,6 +87,19 @@ class ARROW_EXPORT RunEndEncodedArray : public Array {
   /// The physical offset to the array is applied.
   const std::shared_ptr<Array>& values() const { return values_array_; }
 
+  /// \brief Returns an array holding the logical indexes of each run end
+  ///
+  /// If a non-zero logical offset is set, this function allocates a new
+  /// array and rewrites all the run end values to be relative to the logical
+  /// offset and cuts the end of the array to the logical length.
```
[the remark this answers is not in the document]
Comment (felipecrv) [reply cpp/src/arrow/array/array_run_end.h:94]: Pushed a fix.

File: cpp/src/arrow/array/array_run_end.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (felipecrv) [reply cpp/src/arrow/array/array_run_end.cc]: Pushed a commit fixing this.

File: cpp/src/arrow/array/array_run_end.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (felipecrv) [reply cpp/src/arrow/array/array_run_end.cc:108]: Check added. :heavy_check_mark:

**Review by zeroshade — APPROVED** [review]

**Pull request merged**